### PR TITLE
Always generate merge gather for select statement with locking clause…

### PR DESCRIPTION
**Pull request by kainwen**
… and sort clause.

LockRowsPath will clear the pathkeys info since
when some other transactions concurrently update
the same relation then it cannot guarantee the order.
Postgres will not consider parallel path for the
select statement with locking clause (it sets parallel_safe
to false and parallel_workers to 0 in function
create_lockrows_path). However, Greenplum contains many
segments and is innately parallel. If we simply clear
the pathkey then if later we need a gather, we will
not choose merge gather so even if there is no concurrent
transaction, the data is not in order. See Github issue:
https://github.com/greenplum-db/gpdb/issues/9724.
So just before the finaly gather, we save the pathkeys
and then invoke create_lockrows_path. In the following
gather, if we found saved_pathkeys is not NIL, we just
create a merge gather.

Another need to mention here is that, the condition that
code can reach create_lockrows_path is very rigour: the
query has to be a toplevel select statement and the range
table has to be a normal heap table and there is only one
table invole the query and many other conditions (please
refer `checkCanOptSelectLockingClause` for details. As the
above analysis, if the code reaches here and the path->pathkeys
is not NIL, the following gather has to be the final gather.
This is very important because if it is not the final gather,
it might be used by others as subpath, and its pathkeys is
not NIL which breaks the rules for lockrows path. We need
to keep the pathkeys in the final gather here.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by kainwen** [comment]
The pr pipeline fails at

```
======================================================================
DIFF FILE: ../gpdb_src/src/test/regress/regression.diffs
----------------------------------------------------------------------

--- /tmp/build/b235ddbb/gpdb_src/src/test/regress/expected/appendonly.out	2020-03-13 13:31:16.984884470 +0000
+++ /tmp/build/b235ddbb/gpdb_src/src/test/regress/results/appendonly.out	2020-03-13 13:31:17.108884480 +0000
@@ -709,9 +709,9 @@
   SELECT oid FROM pg_class WHERE relname='tenk_ao2'));
    case    | objmod | last_sequence | gp_segment_id 
 -----------+--------+---------------+---------------
+ NormalXid |      0 | 1-2900        |             2
  NormalXid |      0 | >= 3300       |             0
  NormalXid |      0 | >= 3300       |             1
- NormalXid |      0 | >= 3300       |             2
  NormalXid |      1 | zero          |             0
  NormalXid |      1 | zero          |             1
  NormalXid |      1 | zero          |             2
@@ -732,9 +732,9 @@
   SELECT oid FROM pg_class WHERE relname='tenk_ao2'));
    case    | objmod | last_sequence | gp_segment_id 
 -----------+--------+---------------+---------------
+ NormalXid |      0 | 1-2900        |             2
  NormalXid |      0 | >= 3300       |             0
  NormalXid |      0 | >= 3300       |             1
- NormalXid |      0 | >= 3300       |             2
  NormalXid |      1 | zero          |             0
  NormalXid |      1 | zero          |             1
  NormalXid |      1 | zero          |             2

======================================================================
```

It seems a flaky case. The code in this pr has nothing to do with this.

**Comment by hlinnaka** [comment]
So, how will the sorted Gather Motion behave if the input is not in order? That can happen if there are concurrent updates. At least we should add comments in execMotion.c to explain that the inputs might not be in order, after all. There better not be assertions for that, for example.

**Comment by hlinnaka** [comment]
Could we just consider this not a bug? The ordering is always going to "best effort" anyway, because of the possibility of concurrent updates. (Unless the transaction is REPEATABLE READ, I guess)

**Comment by kainwen** [comment]
> So, how will the sorted Gather Motion behave if the input is not in order? That can happen if there are concurrent updates. At least we should add comments in execMotion.c to explain that the inputs might not be in order, after all. There better not be assertions for that, for example.

Good question. I never consider that. Will have a try.

> Could we just consider this not a bug? The ordering is always going to "best effort" anyway, because of the possibility of concurrent updates. (Unless the transaction is REPEATABLE READ, I guess)

I agree that this is not a bug. I believe the correct usage for user to write the the sql code
`select * from t order by limit 1 for update` should be under REPEATABLE READ isolation level, 
that is if there is concurrently update firstly lock the tuples, than this `select for update` transaction
should abort. If the isolation level is read committed, then user should keep in mind that the statement 
can not guarantee order.

So in summary:
   1. I will do some test for merge gather if one channel of data is not in order.
   2. A better idea might be:
      a. if isolation level is read committed, just print a warning to show that order is not guaranteed.
      b. if isolation level is repeatable read, then just do not clear the pathkey?

But the above logic does not happen in Postgres. Not sure if they are going to the correct direction.

**Comment by kainwen** [comment]
> Could we just consider this not a bug? The ordering is always going to "best effort" anyway, because of the possibility of concurrent updates. (Unless the transaction is REPEATABLE READ, I guess)

So A quick test shows that our merge gather does not have any assert on orders. I will go through the code to verify this. Not sure if this is good news.

Next, 

> So in summary:
> I will do some test for merge gather if one channel of data is not in order.
> A better idea might be:
> a. if isolation level is read committed, just print a warning to show that order is not guaranteed.
> b. if isolation level is repeatable read, then just do not clear the pathkey?
> But the above logic does not happen in Postgres. Not sure if they are going to the correct direction.

I think Postgres do not generate parallel sort path and that is the reason it does not do so.
Not sure if it is good to consider ISOLATION level during planner stage.

Any ideas? @hlinnaka 

Thanks!

**Comment by kainwen** [comment]
I go through the code of merge gather in GPDB, IMO, it does not have assert of each segment's data is in order.

So, can we add some comments there, and adopt the method in this PR? 

Any ideas?

Thanks @hlinnaka
